contributors/design-proposals/device-plugin.md

```diff
@@ -0,0 +1,557 @@
+* We however do not expect Kubelet to fail or restart any pods or containers
+  running that are using these devices.
+* If Kubelet is in the process of allocating a device, then it should fail
+  the container process.
+
+If the Kubelet fails or restarts, we expect the Device Plugins to know about
+it through gRPC's Keep alive feature and try to reconnect to Kubelet.
+
+When Kubelet fails or restarts it should know what are the devices that are
+owned by the different containers and be able to rebuild a list of available
+devices.
+We are expecting to implement this through a checkpointing mechanism that Kubelet
+would write and read from.
+
+
+## API Changes
+
+When discovering the devices, Kubelet will be in charge of advertising those
+resources to the API server as part of the kubelet node update current protocol.
+
+We will be using extended resources to schedule, trigger and advertise these
+Devices.
+When a Device plugin registers two `foo-device` the node status will be
+updated to advertise 2 `extensions.kubernetes.io/foo-device`.
+
+If a user wants to trigger the device plugin he only needs to request this
+through the same mechanism as OIRs in his Pod Spec.
+
+*Future*:
+
+In the future (beta) we would like to advertise each device and provide
+introspection capabilities (what pod uses what device).
+
+One solution would be to add a `Device` structure that could be defined
+as follows:
+
+```golang
+// E.g:
+// struct Device {
+//    ResourceName: "nvidia-gpu"
+//    ID: "GPU-fef8089b-4820-abfc-e83e-94318197576e"
+//    State: "Healthy",
+//    Properties: {
+//        "Family": "Pascal",
+//        "Memory": "4G",
+//        "ECC"   : "True",
+//    }
+//}
+type Device struct {
+	ResourceName string
+	ID           string
+	Status       DeviceHealthStatus
+	Properties   map[string]string
+}
+
+// Extend the node status
+type NodeStatus struct {
+	DevCapacity []Device
+}
+```
+
+We also introduce the `Devices` field in the Containers status so that user
+can know what devices were assigned to the container.
+
+```golang
+type ContainerStatus struct {
+	Devices []Device
+}
+```
+
+# Upgrading your cluster
+
+*TLDR:* 
+Given that we cannot guarantee that the Device Plugins are not running
+a daemon providing a critical service to Devices and when stopped will
+crash the running containers, it is up to the vendor to specify the
+upgrading scheme of their device plugin.
+
+However, If you are upgrading either Kubelet or any device plugin the safest way
+is to drain the node of all pods and upgrade.
+
+Depending on what you are upgrading and what changes happened then it
+is completely possible to only restart just Kubelet or just the device plugin.
+
+## Upgrading Kubelet
+
+This assumes that the Device Plugins running on the nodes fully implement the
+protocol and are able to recover from a Kubelet crash.
+
+Then, as long as the Device Plugin API does not change upgrading Kubelet can be done
+seamlessly through a Kubelet restart.
+
```

Comment:
I like the idea if trying to show people how to restart the gRPC service and re-register. 
However, I don't think a package is the right model.
I don't see how it would be feasible without imposing a strict model on how we expect people to start their gRPC service.

Maybe an example would be a better model ?